### PR TITLE
[do not review] chore: add WasmRegistry and WasmModule

### PR DESCRIPTION
**Pull request by kostasrim**

### CODE REVIEW

File: src/server/wasm/wasm_registry.cc [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+  // 2. Setup && compile
+  //  wasmtime_module_t* module = nullptr;
+  //  wasm_byte_vec_t wasm;
+  //  wasm_byte_vec_new_uninitialized(&wasm, is_file_read->size());
+  //  std::memcpy(wasm.data, is_file_read->data(), wasm.size);
+
+  // The following line is problematic, it causes a SEGFAULT when we switch fibers
+  // after the command gets executed
+
+  //  auto error = wasmtime_module_new(engine_, (uint8_t*)wasm.data, wasm.size, &module);
```
Comment (kostasrim) [inline src/server/wasm/wasm_registry.cc:75]: if we comment this out we crash when we switch fibers @dranikpg

File: src/server/wasm/wasm_registry.cc [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+  // There are more configs we should look at. In particular we can spend a little more
+  // time in compiling but generate a more optimized binary as well ass signal handling config.
+  wasi_config_ = wasi_config_new();
+  CHECK(wasi_config_);
+  wasi_config_inherit_argv(wasi_config_);
+  wasi_config_inherit_env(wasi_config_);
+  wasi_config_inherit_stdin(wasi_config_);
+  wasi_config_inherit_stdout(wasi_config_);
+  wasi_config_inherit_stderr(wasi_config_);
```
Comment (dranikpg) [inline src/server/wasm/wasm_registry.cc:36]: is there a way to start it without wasi? We don't need it, it should be a pure sandbox

File: src/server/wasm/wasm_registry.cc (hunk 1 above)
Comment (dranikpg) [reply src/server/wasm/wasm_registry.cc:75]: how can we switch fibers if there are no preemption points? The mutex below? Let's make everything thread-local or use a spinlock

File: src/server/wasm/wasm_registry.h [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  WasmRegistry(WasmRegistry&&) = delete;
+  ~WasmRegistry();
+  std::string Add(std::string_view path);
+  bool Delete(std::string_view name);
+
+  // Very light-weight. Each Module is compiled *once* but each UDF call, e,g, `CALLWASM`
+  // will spawn its own instantiation of the wasm module. This is fine, because the former
+  // is the expensive operation while the later is used as the context upon the function
+  // will execute (effectively allowing concurrent calls to the same wasm module)
+  class WasmModuleInstance {
```
Comment (dranikpg) [inline src/server/wasm/wasm_registry.h:65]: Better leave a comment what the difference between engine/store/module are 🙂

**Review by dranikpg — APPROVED** [review]

File: src/server/wasm/wasmtime.hh [hunk 4]
```diff
@@ -0,0 +1,3268 @@
+/**
```
Comment (kostasrim) [inline src/server/wasm/wasmtime.hh:1]: @dranikpg taken as is

File: src/server/wasm/wasm_registry.h [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  mutable util::fb2::SharedMutex mu_;
+
+  // Global available for all threads
+  // see: https://docs.wasmtime.dev/c-api/wasmtime_8h.html in section thread safety
+  wasmtime::Engine engine_;
+  wasmtime::Linker linker_;
+  wasmtime::Store store_;
+  static wasmtime::Config GetConfig() {
+    wasmtime::Config config;
+    config.epoch_interruption(false);
```
Comment (kostasrim) [inline src/server/wasm/wasm_registry.h:87]: We still have the same issue after we switch fiber we crash. But loading and executing wasm works fine, so for example when I merge the code in `WasmFamily::Load` with `WasmFamily::Call` it properly loads and executes the wasm binary. However it still crashes after it exits just like when we split this in two steps as described above.

File: src/server/wasm/wasm_registry.cc (hunk 1 above)
Comment (kostasrim) [reply src/server/wasm/wasm_registry.cc]: I am not sure, I had also commented out the locks all together. I will play around and ping

File: src/server/wasm/wasm_registry.cc (hunk 2 above)
Comment (kostasrim) [reply src/server/wasm/wasm_registry.cc]: `wasi` is the portable abstraction of a system interface (for example reading from a socket `read`). Some of these are implemented while others have limited functionality. Either way, I think it's fine to have those.

**Review by dranikpg — APPROVED** [review]
Let's merge, I'll steal your branch :neckbeard:

**Pull request merged**